### ISSUE TITLE
Plugin should fail when new versions of a dependency are signed with a different key

### ISSUE THREAD

**Issue by Diagoras** [s4u/pgpverify-maven-plugin#17]
If a signed dependency is updated to a new version signed with a different key, the plugin should fail the build with an appropriate error message since this is a bit strange. This involves keeping track of a mapping of groupId + artifactId to keys, something which is already partially implemented with the "keyMap".

**Comment by Kortanul** [comment]
@Diagoras IMO this already happens, does it not? If you enable the option to fail on a missing signature, then it will fail when the dependency is updated because the signature is missing from the file. the user who adds the new signature to the file should take that as an opportunity to compare the old and new signatures.

**Comment by Diagoras** [comment]
@Kortanul Hmm, its been a while since I looked at this one. As I recall, the plugin didn't keep track of signatures used across different versions of the same artifact. So a new version with a different signature would pass, because all it does is check that the signature exists and is valid - it doesn't check if it's consistent with previous signing keys. Like if SSH only checked that the server presented a key and that the key was properly constructed, but didn't check if it matched the known fingerprint for that server.

**Comment by martinm1000** [comment]
Hello, did anyone consider this idea ?

if in my pom, I could somehow define keys I trust for an artifact ?
This plugin could then easily fail if there is no signature or an unrecognized one.

I would like to start verifying my project dependencies, but without this issue having
a fix I'm not sure how useful the plugin is !

**Comment by Diagoras** [comment]
@martinm1000 That's a reasonable ask, but I think you might want to make a new issue for it.

**Comment by slawekjaranowski** [comment]
Hi, you can prepare file with definition which artifact should be signed by specific key id.

Please see examples in IT tests:
https://github.com/s4u/pgpverify-maven-plugin/tree/master/src/it/sigOkKeysMap
https://github.com/s4u/pgpverify-maven-plugin/tree/master/src/it/sigOkKeysMapMultiModule

**Comment by martinm1000** [comment]
@slawekjaranowski 
Since the key id can be non-unique (and in some format can be computed for collisions) can we specify a key fingerprint ?

**Comment by slawekjaranowski** [comment]
Yes, you can use fingerprint, as I know keyId is  part of fingerprint - it last bytes from it

**Comment by martinm1000** [comment]
@slawekjaranowski 
Ok, so I can specify a key fingerprint that should sign an artifact (and not keyId). Nice.

(Persharp I should open an issue for this question)

My point is that having looked at the code a little, I think the certificates are internally mapped using their KeyId (collision prone) and not their fingerprint (really unique).

Have you tested if the plugin will fail if there is an attack where someone resigned an artifact with a certificate that cloned the real certificate keyid (when you retrieve a certificate using the keyId what if there is more than one certificate returned ?)

This is why I wanted to know if I can at least specify a complete fingerprint and be sure the plugin **won't just use that fingerprint KeyId to map to its certificate**.

in other words, do you think the plugin will behave correctly ? 

refs: 
https://lkml.org/lkml/2016/8/15/445
https://searchsecurity.techtarget.com/answer/How-can-PGP-short-key-IDs-be-protected-from-collision-attacks

**Comment by Diagoras** [comment]
@martinm1000 I agree that using fingerprint instead of key ID is a good idea, but I think you should probably open a new issue for it since it seems otherwise unrelated to this one.

**Comment by slawekjaranowski** [comment]
#42 - covers issue for using fingerprint instead key id.

I close this - I believe that keys map is good enough to resolve this issue.

We have new option to manage artifacts without signature #44
